src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectState.cs

```diff
@@ -198,23 +248,22 @@ public ProjectState WithChangedHostDocument(HostDocument hostDocument, SourceTex
                 throw new ArgumentNullException(nameof(hostDocument));
             }
 
-            if (!Documents.ContainsKey(hostDocument.FilePath))
+            if (!Documents.TryGetValue(hostDocument.FilePath, out var document))
             {
                 return this;
             }
 
-            var documents = new Dictionary<string, DocumentState>(FilePathComparer.Instance);
-            foreach (var kvp in Documents)
-            {
-                documents.Add(kvp.Key, kvp.Value);
-            }
+            var documents = Documents.SetItem(hostDocument.FilePath, document.WithText(sourceText, version));
 
-            if (documents.TryGetValue(hostDocument.FilePath, out var document))
+            if (ImportsToIncludingDocuments.TryGetValue(hostDocument.TargetPath, out var relatedDocuments))
             {
-                documents[hostDocument.FilePath] = document.WithText(sourceText, version);
+                foreach (var relatedDocument in relatedDocuments)
+                {
+                    documents = documents.SetItem(relatedDocument, documents[relatedDocument].WithImportsChange());
+                }
```

Comment:
This is the part of this PR where it all comes together. Now that we know which documents are imports, and which documents are related to which imports, we can reset their state when an import changes.